Nonlinear extrusion

abortz opened the pull request:
### Description

Many have noticed that at high flow rates, extruders will under extrude even when they are otherwise not misbehaving or skipping. Therefore we can simply compensate for that underextrusion (even though the cause of that underextrusion is still being debated). This is an implementation paralleling some existing work in RepRap. This feature boosts extruded length and speed based on the instantaneously commanded extrusion rate. Note that some implementations elsewhere base the correction on the commanded or average segment extruder feedrate -- certainly an easier and faster implementaiton, but this is not sufficient to correct under/overextrusion during acceleration and deacceleration, which I find to be a super valuable feature. This feature is completely independent of linear/pressure advance and is correcting a different nonideality of our extruders.

Note this is a barebones PR and I know there are some minor areas needing further implementation, but I'd like to get feedback on the general strategy here as early as possible.

### Requirements

1. Requires ADAPTIVE_STEP_SMOOTHING and forces the minimum oversampling level to 1.
2. Probably requires more than a minimum power board (but I can confirm it runs well on a Creality board for a Ender 3 S1). 
3. Parameters are not intuitive, and require some method of calibration. I calibrated my setup by measuring filament length at the input to the extruder, which has the added benefit of being easy and theoretically consistent with the correction strategy used here.

### Benefits

Can be used to calibrate actual extruded length to match commanded extruded length regardless of feedrates. The most obvious result of this is getting more correct extrusion widths at all speeds and accelerations, which results in more accurate dimensions for thin walls and better surface finish via correcting overextrusion at the beginning and end of line segments.

### Configurations

To enable define NONLINEAR_EXTRUSION in Configuration_adv.h and set the parameters via M592. The M592 command has been defined to be backwards-compatible with the RepRap version, although there are some subtle differences.

With respect to actual parameters, I can only really suggest doing your own calibration. For my Ender 3 S1 with a bimetal heatbreak, prior to enabling this feature my max usable flow rate was 18mm3/s, with skipping starting shortly beyond that. At that max flow rate, I measured around 5% underextrusion (not much compared to other reports, so it'd be great to get some testing by others who have bigger issues). I measured at a few other flow rates, and ended up with a purely quadratic correction term. So I use the following in my filament start G-code:

M592 A0 B0.001147 C1 ; nonlinear extrusion params @ .4mm nozzle PLA+ 205C

### Related Issues

Requested in #9852 
Also requested by Klipper users in Klipper3d/klipper#2610 and (partially?) implemented in https://github.com/Klipper3d/klipper/pull/4231

abortz committed:
Initial working implementation of nonlinear extrusion

abortz committed:
Support negative params for nonlinear extrusion

abortz commented:
I've collected some more data from testing with my other frequently used materials, ASA and TPU. Overall it seems materials start to underextrude only after a certain feedrate (which should come as no surprise to anyone) and the required boost in extrusion can be modelled very accurately by a quadratic, so I've revised the PR to support negative coefficients and to clamp the boost to a minimum -- namely the boost at a feedrate of 0 mm/s.

In terms of usefulness, it has a minor but noticeable effect on top/bottom surface finish in PLA and ASA, but the big benefit is with TPU, where this correction lets me print at roughly 3x the flow rate I used to be able to. The required nonlinear correction is much, much bigger for TPU (which also should surprise no one). At my extruder's max useable feed rate, the measured flow is 22% low.

thinkyhead commented on Marlin/src/module/planner.cpp:
```diff
@@ -2005,7 +2005,7 @@ bool Planner::_populate_block(
     dm.e = (dist.e > 0);
     const float esteps_float = dist.e * e_factor[extruder];
-    const uint32_t esteps = ABS(esteps_float) + 0.5f;
+    const uint32_t esteps = ABS(esteps_float);
```
This rounding to the nearest number of steps has been in the code since prehistoric times. Are you sure it should be removed?

abortz replied:
I didn't understand why we truncate all of the other axis steps but round the esteps so I removed it. I think you're the one who added it 7 years ago, you'd know better than anyone. I kinda forgot I made this change -- it wasn't really supposed to be part of this PR.

tombrazier replied:
I think this changed should be reverted.

abortz replied:
Okay. Why do we currently round esteps and not other axis?

thinkyhead replied:
I'll keep looking into that history. Meanwhile, I'm not sure. It was likely a considered choice at some point. Maybe it is to to mitigate the use of `e_factor` and the introduction of further `float` imprecision. Maybe to have the nearest magnitude correspondence between `esteps_float` and `esteps`. It's hard to say whether it's of any import without looking at some different situations. I tend to agree that a single half-E-step probably doesn't matter in the grand scale of E-steps.

thinkyhead committed:
apply review changes

thinkyhead committed:
fix test

tombrazier commented:
I have been thinking something like this would be a good idea but did not want to get into working out how to implement it. This is an interesting implementation and I like how simply it fits the existing code.

Several comments:
- As I understand the code, a quadratic function is used to scale the E extrusion rate. I think it is `A * velocity + B * velocity^2 + C`. The ordering of A, B and C does not make sense. It should follow the standard quadratic representation of ax^2 + bx + c.
- Stepper::calc_multistep_timer_interval() is not the right place for modifying `advance_dividend.e`. That will be a very much unexpected side effect of calling the function. This code should have its own function that is called before or after `calc_multistep_timer_interval()` (in all four places this function is called).
- Will multiplying by the quadratic always result in a value for `advance_dividend.e` that is less than `advance_divisor`?
- Does `const uint32_t velocity = ne_scale * step_rate;` in line 2205 not need a `>> 24`?
- It would be great if this could work on slower hardware as well (I have an AVR). How about not using `ADAPTIVE_STEP_SMOOTHING` and instead scale `advance_dividend.e` up in the planner and then use the quadratic function to scale it by a value less <= 1.0 in the stepper module?
- This is going to mess up `count_position.e` values and, therefore, perhaps logic that depends on `Stepper::position()` and maybe logic that uses `Stepper::_set_position()` and `Stepper::set_axis_position()`. I'm not sure whether this is a problem.

PS I am pretty pleased to see this for another reason: you have clearly understood how the stepper logic works and the more people who do, the better I think.

tombrazier commented:
I also really like how this interacts with linear advance. By basing the scaling factor on the intended E speed, but modifying the Bresenham counter (which LA also uses), it scales based on the right number but it also applies the scaling correctly to LA if LA is enabled.

tombrazier commented:
Further thoughts:

- I think `STEP_MULTIPLY()` would probably work for the multiplications and would be a lot faster for AVR. Ranges would need checking though. The first parameter to `STEP_MULTIPLY()` needs to be a 24 bit value and the return value is only 16 bits. The second parameter can be a full 32 bit. The results is the product of the parameters right shifted 24 bits.
- There is one more multiplication by `velocity` than is needed in the logic that calculates `vd`. A better approach (using `STEP_MULTIPLY` and rearranging A, B and C as I suggest) is:
```
  int32_t vd = STEP_MULTIPLY(velocity, nefix.A);
  vd = STEP_MULTIPLY(velocity, vd + nefix.B);
  vd = vd + nefix.C;
```
- For my printer, a quadratic function is not needed. Linear will suffice. Losing all the `velocity^2` logic would be a win. I wonder if it is worth having a Configuration_adv.h option that specifies whether linear or quadratic is used.

tombrazier commented:
This is going to be a great add on for speed benchies!

abortz commented:
Thanks for the comments @tombrazier !

- The ordering of A, B and C is to maintain compatibility with the RepRap implementation of M592. I don't have a strong opinion on how important this is.
- Piggybacking on ``calc_multistep_timer_interval`` is really useful because we need the ``step_rate`` before it gets changed by other features, and it's called precisely when the ``step_rate`` can change. I agree it's weird to put this compensation in there, but it happened to be precisely the right place. Moving it out ends up duplicating code and risks missing a ``step_rate`` change.
- Dividend will be smaller than divisor up to a factor of 2x compensation, as a result of forcing an oversampling level of >=1. There's nothing that deals with >2x compensation -- it might work, or it might just cap at 2x. This isn't great at all but putting code in the time-critical areas to deal with it properly is probably worse. 2x compensation is **a lot** though.
- The straight multiplication to compute velocity is correct, because we are converting from an integer to an 8.24 fixed-point number. It's definitely not clear, but that's mostly because we don't have a real fixed-point math library.
- I like the idea of moving some of the logic to the planner so we can ditch the requirements for adaptive step smoothing and oversampling which are very much shortcuts to getting this working. A very good candidate for an improvement in a future PR.
- Yup, this definitely affects ``count_position``. In one sense this is fine, since the nonlinear compensation **is** changing the number of steps taken. If you really care about the **position** of the E axis wouldn't you want it in mm/inches and therefore get the pre-corrected value from somewhere else?
- Actually, I'm pretty sure this isn't a perfect interaction with linear advance. But I don't actually have a good intuition on the LA mechanism in the first place. In practice it doesn't seem to be an issue, but again my testing has been at fairly moderate compensation multipliers.
- I initially used ``STEP_MULTIPLY`` but I needed *signed* fixed-point math so I just implemented it inline and added the 32-bit requirement. I imagine it'd be possibly to implement for AVR but I didn't want to start on a whole fixed-point abstraction layer...
- It's definitely possible to further optimize the multiplications in the time-critical section, but you do have to watch out for keeping the intermediate values in range of the 8.24 fixed-points.
- I expect limiting it to only a linear coefficient would limit the increase in useable flow rates quite a lot. Are you sure your printer only needs a linear compensation? In my experiments I found the opposite, that a purely quadratic compensation was surprisingly effective. But for lower-power boards I totally understand why we should think about that.

Sadly, I think I only 80% understand the stepper logic, but it was enough to whip up this PR. :-)

abortz committed:
Fix naming

tombrazier commented:
> The ordering of A, B and C is to maintain compatibility with the RepRap implementation of M592.

Oh, well good work on maintaining compatibility. That is important.

I see RRF doesn't actually have the `C` value. [`M221`](https://marlinfw.org/docs/gcode/M221.html) already does this job so it makes sense not to have `C`. In the absence of `C`, I can live with `A` and `B` being reversed.

Reading the documentation on reprap.org, I also see RRF has an upper bound `L`. This would be a helpful addition if we did what I suggested with prescaling in the planner. Then we simply prescale by `1 + L` and in the stepper module reduce `ne_scale` equivalently.

I see from this comment in `M592.cpp`:
```
*  The amount of extrusion is multiplied by max(C, C + A*v + B*v^2) where v is extruder velocity in mm/s.
```

But I don't think the `max` is actually happening. I think it would makes sense to replace the `C` parameter with an `L` parameter and apply it with a call to `_MAX()` which is Marlin's platform independent version of max.

> I agree it's weird to put this compensation in there, but it happened to be precisely the right place. Moving it out ends up duplicating code and risks missing a step_rate change.

You haven't convinced me, not least because I think `calc_multistep_timer_interval()` gets inlined anyway. Nonlinear extrusion is similar to linear advance in being an optional extra that just happens to need to be calculated in all the same places that `calc_multistep_timer_interval()` is called. It makes sense to follow the same pattern and not have an unexpected side effect. Unless we rename `calc_multistep_timer_interval()` to something like `update_step_rate()` and have the function also modify `ticks_nominal` rather than return it to be modified by the calling code.

> Yup, this definitely affects `count_position`...

I suspect this is not a problem. Change count_position for the `X`, `Y` or `Z` axes and that will mess up homing. But we don't home `E`. I would have to trawl through a bunch of code to be sure there aren't any unexpected side-effects, though.

> If you really care about the position of the E axis wouldn't you want it in mm/inches and therefore get the pre-corrected value from somewhere else?

The planner module maintains its own record of current position in mm. That's what `M114` reports. But `M114 D` goes deeper and reports the stepper module value as well and also reports if the two are out of sync. I think one could simply say that it is expected that the two would get out of sync when using nonlinear extrusion.

> Actually, I'm pretty sure this isn't a perfect interaction with linear advance. But I don't actually have a good intuition on the LA mechanism in the first place.

I know it inside out. LA uses the same bresenham counters. It just updates at a different frequency to the rest of the step logic. This allows it to piggy-back on the `STEP_MULTIPLY()` result so that no extra multiplications are needed. The consequence is that when nonlinear extrusion changes `advance_dividend.e`, it affects LA as well. But now that I think about it, your PR does not have the equivalent of the adaptive step smoothing logic for LA. So that will sometimes break LA. Prescaling in the planner, again, could hopefully help here.

> I initially used STEP_MULTIPLY but I needed signed fixed-point math...

Ah, yes I see that now. I can implement a signed version of the multiplication functions is assembly if needed for AVR.

> Are you sure your printer only needs a linear compensation?

Yeah, I have a linear function that works quite well. I am sure quadratic would give a better fit, but for the performance cost, I am happy with just linear.

> Sadly, I think I only 80% understand the stepper logic

That's 80% more than about 8 billion people understand!

I think your suggestion about a follow-up PR makes sense. Since I am the one with the AVR there's a good chance I will submit one.

Before this is merged, I'd really quite like to see the conversation around the `C` and `L` parameters resolved. And breaking LA should probably be addressed - perhaps with SanityCheck.h for now.

abortz commented:
> I see RRF doesn't actually have the `C` value. [`M221`](https://marlinfw.org/docs/gcode/M221.html) already does this job so it makes sense not to have `C`. In the absence of `C`, I can live with `A` and `B` being reversed.

The C param and M221 are subtly different, because the nonlinear compensation only applies on forward extrude moves. I've found this to be useful, since I have my E steps/mm calibrated without a nozzle, and then can apply the required boost on forward moves of ~0mm/s with C (for my setup, ~1-5% over reverse moves)

> Reading the documentation on reprap.org, I also see RRF has an upper bound `L`. This would be a helpful addition if we did what I suggested with prescaling in the planner. Then we simply prescale by `1 + L` and in the stepper module reduce `ne_scale` equivalently.

Yup, perfect addition if/when we do the prescaler refactoring.

> I see from this comment in `M592.cpp`:
> 
> ```
> *  The amount of extrusion is multiplied by max(C, C + A*v + B*v^2) where v is extruder velocity in mm/s.
> ```
> 
> But I don't think the `max` is actually happening. I think it would makes sense to replace the `C` parameter with an `L` parameter and apply it with a call to `_MAX()` which is Marlin's platform independent version of max.

It is happening ``NOLESS(vd, 0)``, and it's rather important. Best fit for a bunch of my material/temp setups includes a negative linear coefficient, which is likely because nonlinear effects don't start until a certain extruder velocity. Think of the compensation as a function like (v - X)^2 + C, where X is the velocity where effects start. But we don't want to **reduce** the extrusion below the v=0 case, hence the clamping.

> You haven't convinced me, not least because I think `calc_multistep_timer_interval()` gets inlined anyway. Nonlinear extrusion is similar to linear advance in being an optional extra that just happens to need to be calculated in all the same places that `calc_multistep_timer_interval()` is called. It makes sense to follow the same pattern and not have an unexpected side effect. Unless we rename `calc_multistep_timer_interval()` to something like `update_step_rate()` and have the function also modify `ticks_nominal` rather than return it to be modified by the calling code.

Fair enough.

> > Actually, I'm pretty sure this isn't a perfect interaction with linear advance. But I don't actually have a good intuition on the LA mechanism in the first place.
> 
> I know it inside out. LA uses the same bresenham counters. It just updates at a different frequency to the rest of the step logic. This allows it to piggy-back on the `STEP_MULTIPLY()` result so that no extra multiplications are needed. The consequence is that when nonlinear extrusion changes `advance_dividend.e`, it affects LA as well. But now that I think about it, your PR does not have the equivalent of the adaptive step smoothing logic for LA. So that will sometimes break LA. Prescaling in the planner, again, could hopefully help here.

I haven't changed the adaptive step smoothing logic, just forced 1 level of oversampling. So LA shouldn't ever be **broken**, but I do believe that the exact value of the LA steps should be a function of velocity when we're doing nonlinear compensation, and I don't believe it is recalculated from it at the moment.

abortz committed:
Move nonlinear calc out of calc_multistep()

tombrazier commented:
> The C param and M221 are subtly different...

Got it.

> It is happening NOLESS(vd, 0), and it's rather important...

Okay, I see that now too.

> I haven't changed the adaptive step smoothing logic, just forced 1 level of oversampling. So LA shouldn't ever be broken, but I do believe that the exact value of the LA steps should be a function of velocity when we're doing nonlinear compensation, and I don't believe it is recalculated from it at the moment.

Actually my thinking was incorrect here as I had forgotten that `advance_isr()` uses `la_dividend` in place of `advance_dividend.e`. So I now think LA is just completely unaffected by this PR. I guess that's what you mean. Ideally `la_dividend` would be scaled by the same factor as `advance_dividend.e`. Another thing for the follow up PR, perhaps.

tombrazier commented:
Anyway, sorry to give you a hard time. I like this feature and am glad you've submitted this PR.

abortz commented:
Ha, no problem @tombrazier! They were good comments.

tombrazier commented:
> I have my E steps/mm calibrated without a nozzle, and then can apply the required boost on forward moves of ~0mm/s with C (for my setup, ~1-5% over reverse moves)

I've been thinking more about this. I'm curious why you need more forward steps than reverse steps when E speed is close to zero. Nozzle pressure should be very low in this case so why the asymmetry?

abortz commented:
> > I have my E steps/mm calibrated without a nozzle, and then can apply the required boost on forward moves of ~0mm/s with C (for my setup, ~1-5% over reverse moves)
> 
> I've been thinking more about this. I'm curious why you need more forward steps than reverse steps when E speed is close to zero. Nozzle pressure should be very low in this case so why the asymmetry?

I honestly don't know. Sliding friction within the nozzle? Viscous losses? Just pressure? There must still be some pressure even at very low extrusion rates, since we're pushing the 1.75mm diameter filament down to the nozzle size. For me it ranges from 0.9% for relatively high-temp (260F) ASA to 5.1% for 90A TPU (215F). My "zero" extrusion rate standard has been 0.5mm/s, as my patience is limited ;-). There is still the unknown mechanism for how we can have underextrusion without skipping in the first place, probably related to gears slipping on the filament, or the filament compressing length-wise.

I'm not super surprised no one else has observed this. It requires testing E steps without a nozzle, which most people don't do, because why would they? This effect is mostly compensated out with slicer flow and retraction settings, with the residual being a very slight net underextrusion proportional to the number of retractions.

thinkyhead committed:
small tweaks. restore DWIN_LCD_PROUI stuff

thinkyhead committed:
Merge branch 'bugfix-2.1.x' into pr/26127

thinkyhead reviewed (approved):
Everything looks great for a 1.0 version. Good job, and thanks for your patience!

GenXEngineer commented:
> There is still the unknown mechanism for how we can have underextrusion without skipping in the first place, probably related to gears slipping on the filament, or the filament compressing length-wise.
> 
> I'm not super surprised no one else has observed this. It requires testing E steps without a nozzle, which most people don't do,

Yep that phenomenon is a trip for sure. Even newby's like me notice and wonder. :)

vovodroid commented:
Hi Andrew, I'm working on enabling SMOOTH_LIN_ADVANCE with NONLINEAR_EXTRUSION.

Any reasons why two NONLINEAR_EXTRUSION blocks are not continuous and separated by other code? I didn't find interaction between those parts.

```
      #if ENABLED(NONLINEAR_EXTRUSION)
        ne_edividend = advance_dividend.e;
        const float scale = (float(ne_edividend) / advance_divisor) * planner.mm_per_step[E_AXIS_N(current_block->extruder)];
        ne_scale = (1L << 24) * scale;
        if (current_block->direction_bits.e && ANY_AXIS_MOVES(current_block)) {
          ne_fix.A = (1L << 24) * ne.A;
          ne_fix.B = (1L << 24) * ne.B;
          ne_fix.C = (1L << 24) * ne.C;
        }
        else {
          ne_fix.A = ne_fix.B = 0;
          ne_fix.C = (1L << 24);
        }
      #endif

      // Calculate the initial timer interval
      interval = calc_multistep_timer_interval(current_block->initial_rate << oversampling_factor);
      // Initialize ac/deceleration time as if half the time passed.
      acceleration_time = deceleration_time = interval / 2;

      #if ENABLED(NONLINEAR_EXTRUSION)
        calc_nonlinear_e(current_block->initial_rate << oversampling_factor);
      #endif
```

abortz commented:
> Hi Andrew, I'm working on enabling SMOOTH_LIN_ADVANCE with NONLINEAR_EXTRUSION.
> 
> Any reasons why two NONLINEAR_EXTRUSION blocks are not continuous and separated by other code? I didn't find interaction between those parts.

You're not wrong, there is no interaction -- it's an artifact of refactoring the nonlinear extrusion code out of calc_multistep_timer_interval at some point. calc_nonlinear_e is called in multiple places, so this is the only one adjacent to another ifdef block. It's been awhile but I suspect I liked the aesthetics of calling calc_nonlinear_e after calc_multistep_timer_interval better. :-)
